ducktape/cluster/remoteaccount.py

```diff
@@ -159,29 +159,55 @@ def _log(self, level, msg, *args, **kwargs):
         msg = "%s: %s" % (str(self), msg)
         self.logger.log(level, msg, *args, **kwargs)
 
-    @property
-    def ssh_client(self):
-        if not self._ssh_client:
-            client = SSHClient()
-            client.set_missing_host_key_policy(IgnoreMissingHostKeyPolicy())
+    def _set_ssh_client(self):
+        client = SSHClient()
+        client.set_missing_host_key_policy(IgnoreMissingHostKeyPolicy())
+
+        self._log(logging.DEBUG, "ssh_config: %s" % str(self.ssh_config))
+
+        client.connect(
+            hostname=self.ssh_config.hostname,
+            port=self.ssh_config.port,
+            username=self.ssh_config.user,
+            password=self.ssh_config.password,
+            key_filename=self.ssh_config.identityfile,
+            look_for_keys=False)
 
-            self._log(logging.DEBUG, "ssh_config: %s" % str(self.ssh_config))
+        if self._ssh_client:
+            self._ssh_client.close()
+        self._ssh_client = client
 
-            client.connect(
-                hostname=self.ssh_config.hostname,
-                port=self.ssh_config.port,
-                username=self.ssh_config.user,
-                password=self.ssh_config.password,
-                key_filename=self.ssh_config.identityfile,
-                look_for_keys=False)
-            self._ssh_client = client
+    @property
+    def ssh_client(self):
+        if (self._ssh_client and
+            self._ssh_client.get_transport() and
+            self._ssh_client.get_transport().is_active()):
+            try:
+                transport = self._ssh_client.get_transport()
+                transport.send_ignore()
+            except Exception as e:
+                self._log(logging.DEBUG, "exception getting ssh_client (creating new client): %s" % str(e))
+                self._set_ssh_client()
+        else:
+            self._set_ssh_client()
 
         return self._ssh_client
 
+    def _set_sftp_client(self):
+        if self._sftp_client:
+            self._sftp_client.close()
+        self._sftp_client = self.ssh_client.open_sftp()
+
     @property
     def sftp_client(self):
-        if not self._sftp_client:
-            self._sftp_client = self.ssh_client.open_sftp()
+        if self._sftp_client:
+            try:
+                self._sftp_client.listdir(".")
```

Comment:
i don't think we want to do this much work to validate this. would the channel used for ftp time out independently (and earlier) than the underlying ssh transport? since this client is layered on top of the `ssh_client`, i think we could do the lighter weight check on just the ssh client. i.e. instead of trying to do something with sftp client, just do the `send_ignore` check and only replace `self._sftp_client` if it that check fails.

if we *do* need to do something with the sftp client itself, maybe do a `stat` instead of a `listdir`. the `stat` should be small and fixed size. `listdir` when cwd is in the wrong place could be quite expensive.

Reply:
I think you're right. we should just need to check on the `self.ssh_client` and update the `self._sftp_client` when that fails

if not, will use `stat` instead